cmd/zb/perf.go

```diff
@@ -622,6 +625,215 @@ func PushChunkStreamed(workdir, url, auth, trepo string, requests int,
+		SetHeader("Content-Type", "application/octet-stream").SetBody(fhandle).Put(loc)
+
+	if err != nil {
+		return err
+	}
+
+	// request specific check
+	statusCode = resp.StatusCode()
+	if statusCode != http.StatusCreated {
+		return errors.ErrUnknownCode
+	}
+
+	// create a manifest
+	manifest := ispec.Manifest{
+		Versioned: imeta.Versioned{
+			SchemaVersion: defaultSchemaVersion,
+		},
+		Config: ispec.Descriptor{
+			Digest: digest,
+			Size:   int64(size),
+		},
+		Layers: []ispec.Descriptor{
+			{
+				MediaType: "application/vnd.oci.image.layer.v1.tar",
+				Digest:    digest,
+				Size:      int64(size),
+			},
+		},
+	}
+
+	content, err := json.Marshal(manifest)
+	if err != nil {
+		return err
+	}
+
+	digest = godigest.FromBytes(content)
+
+	resp, err = resty.R().SetHeader("Content-Type", "application/vnd.oci.image.manifest.v1+json").
+		SetBody(content).Put(fmt.Sprintf("%s/v2/%s/manifests/%s", url, repo, digest.String()))
+
+	if err != nil {
+		return err
+	}
+
+	// request specific check
+	statusCode = resp.StatusCode()
+	if statusCode != http.StatusCreated {
+		return errors.ErrUnknownCode
+	}
+
+	manifestLoc := location(url, resp)
+
+	// download image
+	for count := 0; count < requests; count++ {
+		func() {
+			start := time.Now()
+
+			var isConnFail, isErr bool
+
+			var statusCode int
+
+			var latency time.Duration
+
+			defer func() {
+				// send a stats record
+				statsCh <- statsRecord{
+					latency:    latency,
+					statusCode: statusCode,
+					isConnFail: isConnFail,
+					isErr:      isErr,
+				}
+			}()
+
+			// send request and get response the manifest
+			resp, err := client.R().Get(manifestLoc)
+			if err != nil {
+				isConnFail = true
+
+				return
+			}
+
+			// request specific check
+			statusCode = resp.StatusCode()
+			if statusCode != http.StatusOK {
+				isErr = true
+
+				return
+			}
+
+			manifestBody := resp.Body()
+
+			// file copy simulation
+			_, err = io.Copy(ioutil.Discard, bytes.NewReader(manifestBody))
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			var pulledManifest ispec.Manifest
+
+			err = json.Unmarshal(manifestBody, &pulledManifest)
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			// download blobs
+			for _, layer := range pulledManifest.Layers {
+				blobDigest := layer.Digest
+
+				// send request and get response the blob
+				resp, err := client.R().
+					Get(fmt.Sprintf("%s/v2/%s/blobs/%s", url, repo, blobDigest))
+				if err != nil {
+					isConnFail = true
+
+					return
+				}
+
+				// request specific check
+				statusCode = resp.StatusCode()
+				if statusCode != http.StatusOK {
+					isErr = true
+
+					return
+				}
+
+				blobBody := resp.Body()
+
+				// file copy simulation
+				_, err = io.Copy(ioutil.Discard, bytes.NewReader(blobBody))
+				if err != nil {
+					log.Fatal(err)
+				}
+			}
+
+			latency = time.Since(start)
+		}()
+	}
+
+	resp, err = client.R().Delete((fmt.Sprintf("%s/v2/%s/", url, repo)))
```

Comment:
There is no DELETE of a top-level repo. You have to iterate over all the tags and delete each manifest entry.

https://github.com/opencontainers/distribution-spec/blob/main/spec.md#deleting-tags